File: models/Person.php
<?php

namespace datagutten\InducksORM\models;

use DateTimeImmutable;
use Doctrine\ORM\Mapping as ORM;
use Doctrine\ORM\PersistentCollection;
use Exception;


/**
 * Person
 *
 *
 * @author dromanin
 */
#[ORM\Table(name: 'inducks_person')]
#[ORM\Entity(readOnly: true)]
class Person
{
    #[ORM\Column(type: 'string')]
    #[ORM\Id]
    private string $personcode;

    #[ORM\Column(type: 'string')]
    private string $nationalitycountrycode;

    #[ORM\ManyToOne(targetEntity: Country::class)]
    #[ORM\JoinColumn(name: 'nationalitycountrycode', referencedColumnName: 'countrycode')]
    private Country $nationality;

    #[ORM\Column(type: 'string')]
    private string $fullname;

    #[ORM\Column(type: 'string')]
    private string $official;

    #[ORM\Column(type: 'string')]
    private string $personcomment;

    #[ORM\Column(type: 'string')]
    private string $unknownstudiomember;

    #[ORM\Column(type: 'string')]
    private string $isfake;

    #[ORM\Column(type: 'integer')]
    private int $numberofindexedissues;

    #[ORM\Column(type: 'string')]
    private string $birthname;

    #[ORM\Column(type: 'string')]
    private string $borndate;

    #[ORM\Column(type: 'string')]
    private string $bornplace;

    #[ORM\Column(type: 'string')]
    private string $deceaseddate;

    #[ORM\Column(type: 'string')]
    private string $deceasedplace;

    #[ORM\Column(type: 'string')]
    private string $education;

    #[ORM\Column(type: 'string')]
    private string $moviestext;

    #[ORM\Column(type: 'string')]
    private string $comicstext;

    #[ORM\Column(type: 'string')]
    private string $othertext;

    #[ORM\Column(type: 'string')]
    private string $photofilename;

    #[ORM\Column(type: 'string')]
    private string $photocomment;

    #[ORM\Column(type: 'string')]
    private string $photosource;

    #[ORM\Column(type: 'string')]
    private string $personrefs;

    #[ORM\OneToMany(mappedBy: 'person', targetEntity: EntryJob::class)]
    private PersistentCollection $entryJobs;

    #[ORM\OneToMany(mappedBy: 'person', targetEntity: StoryJob::class)]
    private PersistentCollection $storyJobs;

    #[ORM\OneToMany(mappedBy: 'person', targetEntity: IssueJob::class)]
    private PersistentCollection $issueJobs;

    public function getPersoncode(): string
    {
        return $this->personcode;
    }

    public function getNationalityCountryCode(): string
    {
        return $this->nationalitycountrycode;
    }

    public function getNationality(): Country
    {
        return $this->nationality;
    }

    public function getFullname(): string
    {
        return $this->fullname;
    }

    /**
     * @return PersistentCollection
     */
    public function getEntries(): PersistentCollection
    {
        return $this->entryJobs;
    }

    /**
     * @return PersistentCollection
     */
    public function getStoryJobs(): PersistentCollection
    {
        return $this->storyJobs;
    }

    public function getIssueJobs(): PersistentCollection
    {
        return $this->issueJobs;
    }

    public function getOfficial(): bool
    {
        return $this->official == 'Y';
    }

    /**
     * @return string
     */
    public function getPersonComment(): string
    {
        return $this->personcomment;
    }

    public function getUnknownStudioMember(): bool
    {
        return $this->unknownstudiomember == 'Y';
    }

    public function getIsFake(): bool
    {
        return $this->isfake == 'Y';
    }

    public function getNumberOfIndexedIssues(): int
    {
        return $this->numberofindexedissues;
    }

    public function getBirthName(): string
    {
        return $this->birthname;
    }

    public function getBornDate(): ?DateTimeImmutable
    {
        if (empty($this->borndate))
            return null;
        try
        {
            return new DateTimeImmutable($this->borndate);
        }
        catch (Exception)
        {
            return null;
        }
    }

    public function getBornPlace(): string
    {
        return $this->bornplace;
    }

    public function getDeceasedDate(): ?DateTimeImmutable
    {
        try
        {
            return new DateTimeImmutable($this->deceaseddate);
        }
        catch (Exception)
        {
            return null;
        }
    }

    public function getDeceasedPlace(): string
    {
        return $this->deceasedplace;
    }

    /**
     * @return string
     */
    public function getEducation(): string
    {
        return $this->education;
    }

    /**
     * @return string
     */
    public function getMoviesText(): string
    {
        return $this->moviestext;
    }

    /**
     * @return string
     */
    public function getComicsText(): string
    {
        return $this->comicstext;
    }

    /**
     * @return string
     */
    public function getOtherText(): string
    {
        return $this->othertext;
    }

    /**
     * @return string
     */
    public function getPhotoFilename(): string
    {
        return $this->photofilename;
    }

    /**
     * Photo comment
     * @return string
     */
    public function getPhotoComment(): string
    {
        return $this->photocomment;
    }

    /**
     * @return string
     */
    public function getPhotoSource(): string
    {
        return $this->photosource;
    }

    /**
     * @return string
     */
    public function getPersonRefs(): string
    {
        return $this->personrefs;
    }

    public function __toString(): string
    {
        return $this->fullname;
    }
}

File: inducks/models/person.py
from django.db import models
from inducks.models import helpers


class Person(models.Model):
    personcode = models.CharField(max_length=79, primary_key=True)
    nationality = helpers.InducksForeignKey('Country', related_name='persons',
                                            db_column='nationalitycountrycode',
                                            isv_field='nationalitycountrycode', blank=True, null=True)
    fullname = models.TextField(blank=True, null=True)
    official = helpers.InducksBooleanField()
    personcomment = models.CharField(max_length=221, blank=True, null=True)
    unknownstudiomember = helpers.InducksBooleanField()
    isfake = helpers.InducksBooleanField()
    numberofindexedissues = models.IntegerField(blank=True, null=True)
    birthname = models.TextField(blank=True, null=True)
    borndate = helpers.InducksDateField(max_length=10, blank=True, null=True)
    bornplace = models.CharField(max_length=30, blank=True, null=True)
    deceaseddate = helpers.InducksDateField(blank=True, null=True)
    deceasedplace = models.CharField(max_length=31, blank=True, null=True)
    education = models.CharField(max_length=189, blank=True, null=True)
    moviestext = models.CharField(max_length=879, blank=True, null=True)
    comicstext = models.CharField(max_length=927, blank=True, null=True)
    othertext = models.CharField(max_length=307, blank=True, null=True)
    photofilename = models.CharField(max_length=32, blank=True, null=True)
    photocomment = models.CharField(max_length=68, blank=True, null=True)
    photosource = models.CharField(max_length=67, blank=True, null=True)
    personrefs = models.CharField(max_length=179, blank=True, null=True)

    class Meta:
        db_table = 'inducks_person'

    @property
    def image_url(self):
        return 'https://inducks.org/creators/photos/%s' % self.photofilename

    def __str__(self):
        return self.fullname


class PersonAlias(models.Model):
    person = helpers.InducksForeignKey('Person', related_name='aliases')
    surname = models.CharField(max_length=48, blank=True, null=True)
    givenname = models.CharField(max_length=31, blank=True, null=True)
    official = helpers.InducksBooleanField()

    class Meta:
        db_table = 'inducks_personalias'
        unique_together = ['person', 'surname', 'givenname']
